submissions/zalevskyi/friends-app/app.js

```diff
@@ -0,0 +1,187 @@
+  fetchUserData(connectToServer)
+}
+function fetchUserData(onError) {
+  let apiUrl = 'https://randomuser.me/api/1.3/'
+  let fields = ['name', 'gender', 'location', 'dob', 'phone', 'picture']
+  fetch(`${apiUrl}?results=${USERS_QUANTITY}&inc=${fields.join(',')}`, {
+    headers: { 'Content-Type': 'application/json' },
+  })
+    .then((response) => {
+      if (!response.ok) {
+        throw Error(response.statusText)
+      }
+      return response
+    })
+    .then((response) => response.json())
+    .then((responseJSON) => {
+      pushUserData(responseJSON.results)
+      showResults()
+    })
+    .catch((error) => onError(error))
+}
+function pushUserData(randomUserResults) {
+  randomUserResults.forEach((user) =>
+    USERS_DATA.push({
+      name: user.name.first,
+      surname: user.name.last,
+      gender: user.gender,
+      age: user.dob.age,
+      country: user.location.country,
+      city: user.location.city,
+      phone: user.phone,
+      image: user.picture.large,
+    })
+  )
+}
+function showResults() {
+  RESULTS.innerHTML = getUserDataToShow(USERS_DATA, REQUEST_PARAMETERS)
+    .map((user) => getUserCard(user))
+    .join('')
+}
+function getUserDataToShow(userData, parameters) {
+  if (parameters.sort) {
+    return sort(filter(userData, parameters), parameters.sort, parameters.order)
+  } else {
+    return filter(userData, parameters)
+  }
+}
+function sort(userData, fieldToSort, order) {
+  if (fieldToSort === 'name') {
+    return sortFullName(userData, 'name', order)
+  } else if (fieldToSort === 'surname') {
+    return sortFullName(userData, 'surname', order)
+  } else if (fieldToSort === 'age') {
+    return sortNumeric(userData, 'age', order)
+  } else {
+    return sortString(userData, fieldToSort, order)
+  }
+}
+function sortFullName(userData, firstToSort, order) {
+  let sortedData = userData.slice()
+  let secondToSort = 'surname'
+  if (firstToSort === 'surname') {
+    secondToSort = 'name'
+  }
+  if (order === 'ascending') {
+    orderValue = 1
+  } else {
+    orderValue = -1
+  }
+  sortedData.sort(
+    (a, b) =>
+      `${a[firstToSort]} ${a[secondToSort]}`.localeCompare(
+        `${b[firstToSort]} ${b[secondToSort]}`
+      ) * orderValue
+  )
+  return sortedData
+}
+function sortNumeric(userData, propertyToSort, order) {
+  let sortedData = userData.slice()
+  if (order === 'ascending') {
+    sortedData.sort((a, b) => a[propertyToSort] - b[propertyToSort])
+  } else {
+    sortedData.sort((a, b) => b[propertyToSort] - a[propertyToSort])
+  }
+  return sortedData
+}
+function sortString(userData, propertyToSort, order) {
+  let sortedData = userData.slice()
+  if (order === 'ascending') {
+    orderValue = 1
+  } else {
+    orderValue = -1
+  }
+  sortedData.sort((a, b) => {
+    if (a[propertyToSort] > b[propertyToSort]) {
+      return 1 * orderValue
+    } else if (a[propertyToSort] < b[propertyToSort]) {
+      return -1 * orderValue
+    } else {
+      return 0
+    }
```

Comment:
Looks like overengineering. You can have only `sortByString` function and then pass arguments in the different order.

Please, check merged PRs

Reply:
Thank you for the suggestion. It helped me to think over it once again and improve the code.
But I disagree with your request in several points:
1. Bad naming request. 

> You can have only `sortByString` function and then pass arguments in the different order

I have checked the [last merged PR](https://github.com/kottans/frontend-2021-homeworks/pull/455/files) and found this sample `sortedArr.sort((a, b) => sortByName(a, b))`
Well, in English it literally means: "I want to sort array by **_sorting_** age" while it should be "I want to sort array by **_comparing_** age". You can not sort by sorting, you can sort by comparing different values
Due to this naming (`sortByString`) It was very hard for me to understand what does it mean:

> pass arguments in the different order

Sorting by its meaning should not depend on argument order, as order is output of sorting, but not input. While comparing does depends on order.

2. No clear reason for such comparison function as it is already built-in into language

Again in the [same merged PR](https://github.com/kottans/frontend-2021-homeworks/pull/455/files) I found such comparison function
```
const sortByName = (a, b) => {
  return a.name.first.localeCompare(b.name.first);
}
```
But it adds nothing to built-in function, just:
-  renames it (and does it in bad way by naming sort what is actually compare)
-  changes the way String instance method call to static-like call

3.  Error prone

As you recommended for other code snippet:
`sortAscendingFullName(userData, 'name', 'surname')`
> 'surname', 'name' - don't use this approach. It is JS and you can easily misprint name of the property.

As for me it is even more easily to misprint just one letter or order of letters in multiple places
Just compare my new [implementation with flipping sign of compare result](https://gist.github.com/zalevskyi/6dfea994f6e5ea8401d49cf7984f0d46) vs. the same [function but changing arguments order in compare functions as suggested](https://gist.github.com/zalevskyi/efd923f3fc0cbd28dda26c921a18b936)
I don't think that it is overengineering. There is no space for error of misprinting order of `a` or `b`. There is even no need to think about it and check it. Instead it is clearly stated: flip compare result if needed by `orderValue`.

Reply:
Sorry for being confusing :)

The main idea was next:

You don't need things like `sortAscendingFullName` and `sortDescendingFullName`, because you can have only one function, which will sort by name. For this, you can concatenate the name and full name into one value somewhere.

Then, you don't need `sortAscending` and `sortDescending` functions.

> ⚠️ Please, consider next more like pseudocode and not like the source of copy-paste. You should find a better solution on your own.

> Regarding naming - feel free to use `compareNames` and `compareSurnames` if you like.

It is much simpler code when you will write something like this:

```jsx
const compareAge = (firstFriend, secondFriend) => {
  return firstFriend.dob.age - secondFriend.dob.age;
};
```

Then, also as an example, you can have:

```jsx
const ageSorters = {
  descending: () => {
    friends.sort((a, b) => compareAge(a, b));
  },
  ascending: () => {
    friends.sort(compareAge);
  }
};
```

And use it like this:

```jsx
myForm.addEventListener("change", ({ target: radioButton }) => {
  if (radioButton.name === AGE) {
    ageSorters[radioButton.value]();
  }

  if (radioButton.name === NAME) {
    nameSorters[radioButton.value]();
  }
});
```

Which, possibly, can be simplified to:

```jsx
myForm.addEventListener("change", ({ target: radioButton }) => {
  const sorter =
    radioButton.name === AGE
      ? ageSorters[radioButton.value]
      : nameSorters[radioButton.value];

  sorter();
});
```

And that's all :)

1. You have separated the logic of sorting into two separate domains - sorting by age and sorting by name and moved everything into separately mapped handlers. So you don't need to write endless `if ... else` checks. Instead, you are using declarative code.
1. Because of this you don't have a place for "misprinting"
1. You don't pass any strings as arguments anywhere, which is antipattern in general. And with object keys in particular.

Now compare it with code from your examples.

> Actually it is rather under engineering. Potentially in case of very frequent comparison in extremely data savvy application flipping sing of comparison (multiplying by 1 or -1) will negatively impact performance to some extent and cost money.

If I got a penny every time... just read about premature optimization. As an example [here](https://wiki.c2.com/?PrematureOptimization)

TLDR:

The following set of heuristics is a correction:
1. Make it work.
1. Make it right - the code is readable and every idea is expressed.
1. Make everything work.
1. Make everything right.
1. Use the system and find performance bottlenecks.
1. Use a profiler in those bottlenecks to determine what needs to be optimized. See ProfileBeforeOptimizing.
1. Make it fast. You maintained unit tests, right? Then you can refactor the code mercilessly in order to improve the performance.

Reply:
Thank you very much! It was really helpful to understand.
Effectiveness of simply changing order of function arguments, possibility to use functions as object properties and antipattern of strings as arguments was not obvious to me.
Thank you for the article about optimization. That is going to be next step for me.